internal/yunikorn/sync.go

```diff
@@ -94,55 +95,117 @@ func (s *Service) upsertPartitions(ctx context.Context) ([]*dao.PartitionInfo, e
+// syncQueues fetches queues for each partition and upserts them into the database
+func (s *Service) syncQueues(ctx context.Context, partitions []*dao.PartitionInfo) ([]*dao.PartitionQueueDAOInfo, error) {
 	logger := log.FromContext(ctx)
 
-	// Create a wait group as a separate goroutine will be spawned for each partition
-	wg := sync.WaitGroup{}
-	wg.Add(len(partitions))
+	var errs []error
+	var allQueues []*dao.PartitionQueueDAOInfo
+	var mu sync.Mutex
+	var wg sync.WaitGroup
 
-	// Protect the queues and errors slices with a mutex so multiple goroutines can safely append queues
-	mutex := sync.Mutex{}
+	addErr := func(err error) {
+		mu.Lock()
+		defer mu.Unlock()
+		errs = append(errs, err)
+	}
 
-	var errs []error
-	var queues []*dao.PartitionQueueDAOInfo
+	addQueues := func(queues []*dao.PartitionQueueDAOInfo) {
+		mu.Lock()
+		defer mu.Unlock()
+		allQueues = append(allQueues, queues...)
+	}
+
+	addJob := func(partition *dao.PartitionInfo) {
+		wg.Add(1)
+		jobName := fmt.Sprintf("sync_queues_for_partition_%s", partition.Name)
+
+		err := s.workqueue.Add(func(ctx context.Context) error {
+			defer wg.Done()
+
+			queues, err := s.syncPartitionQueues(ctx, partition)
+			if err != nil {
+				addErr(fmt.Errorf("error syncing partition queues for partition %s: %w", partition.Name, err))
+			} else {
+				addQueues(queues)
+			}
+			return nil
+		}, workqueue.WithJobName(jobName))
 
-	processPartition := func(p *dao.PartitionInfo) {
-		defer wg.Done()
-		queue, err := s.client.GetPartitionQueues(ctx, p.Name)
-		mutex.Lock()
-		defer mutex.Unlock()
 		if err != nil {
-			errs = append(errs, fmt.Errorf("could not get queues for partition %s: %v", p.Name, err))
-		} else {
-			queues = append(queues, queue)
+			logger.Errorf("could not add sync queues job for partition %s: %v", partition.Name, err)
 		}
 	}
 
-	for _, p := range partitions {
-		go processPartition(p)
+	for _, partition := range partitions {
+		addJob(partition)
 	}
 
-	// wait for all partitions to be processed
 	wg.Wait()
 
 	if len(errs) > 0 {
-		return nil, fmt.Errorf("failed to get queues for some partitions: %v", errs)
+		return nil, fmt.Errorf("some errors encountered while syncing queues: %v", errs)
 	}
 
-	queues = flattenQueues(queues)
-	err := s.workqueue.Add(func(ctx context.Context) error {
-		logger.Infow("upserting queues", "count", len(queues))
-		return s.repo.UpsertQueues(ctx, queues)
-	}, workqueue.WithJobName("upsert_queues"))
+	return allQueues, nil
+}
+
+func (s *Service) syncPartitionQueues(ctx context.Context, partition *dao.PartitionInfo) ([]*dao.PartitionQueueDAOInfo, error) {
+	// Fetch partition queues from the YuniKorn API
+	queue, err := s.client.GetPartitionQueues(ctx, partition.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve queues from YuniKorn API: %w", err)
+	}
+
+	// Attempt to update the queue; if it fails, try adding it instead
+	if err := s.repo.UpdateQueue(ctx, queue); err != nil {
```

Comment:
I'm not sure why we don't simply use upsert, but let's leave it for later

Reply:
we used to use `upsertQueue` method. But the idea of https://github.com/G-Research/yunikorn-history-server/issues/143 was to separate out CRUD and move away from upsert. I think we have parent-child relationship between queues which is very difficult to update using the upsert.